README.md
pyoath
======

[![Build Status](https://travis-ci.org/markgollnick/pyoath.svg?branch=master)](https://travis-ci.org/markgollnick/pyoath)
[![Coverage Status](https://img.shields.io/coveralls/markgollnick/pyoath.svg)](https://coveralls.io/r/markgollnick/pyoath?branch=master)

![Two-Factor Authentication](https://raw.githubusercontent.com/markgollnick/pyoath/master/2fa-padlocks.png)

A Python OATH (One-time Authentication) implementation.

Pyoath implements HOTP as defined in RFC 4226, published December, 2005, and
TOTP as defined in RFC 6238, published May 2011, and has been designed to be
used on both the client- and server-sides of two-factor authentication systems.


Requirements
------------

* Python >= 2.6, 2.7, 3.2, 3.3, 3.4…


Installation
------------

**For Users:**

    pip install git+ssh://git@github.com/markgollnick/pyoath@v1.0.0#egg=pyoath-1.0.0

**For Developers:**

    git clone git@github.com:markgollnick/pyoath.git
    cd pyoath
    python setup.py build install
    # Alternatively...
    python setup.py sdist
    pip install dist/pyoath-1.0.0.tar.gz


Usage
-----

Once installed, you can use it as a script (that is, on the client-side)…

    $ pyoath ~/.secrets/secret.key
    123456
    $ pyoath ~/.secrets/secret.key --loop
    
    Authenticator Started!
    :----------------------------:--------:
    :       Code Wait Time       :  Code  :
    :----------------------------:--------:
    +++++++++++++++++++++++++++++: 123456 :
    ....................+++++++++: 234567 :
    .............................: 345678 :
    .........^C
    $ 

…or, you can use it as a library (that is, on the server-side):

    >>> import pyoath
    >>> pyoath.HOTP(b'secret', 0)
    '814628'
    >>> pyoath.HOTP(b'secret', 1, Digit=8)
    '28533881'
    >>> pyoath.TOTP(b'secret')
    '123456'
    >>> pyoath.TOTP(b'secret', Digit=8)
    '12345678'
    >>> import hashlib
    >>> pyoath.TOTP(b'secret', Digit=8, Mode=hashlib.sha512)
    '87654321'


Acknowledgments
---------------

- Special thanks to OpenSSH for the bold notice about poor file access bits:
  <http://www.openssh.com/>
- Special thanks to James Cuff for the
  [Java-based Google Authenticator Desktop Client][2], which inspired this
  project.
- Special thanks to [AJ][] for [padlock][open] [icons][closed].

[1]: http://blog.jcuff.net/2011/02/cli-java-based-google-authenticator.html
[AJ]: https://openclipart.org/user-detail/AJ
[open]: https://openclipart.org/detail/33553/open-padlock-by-anonymous
[closed]: https://openclipart.org/detail/17931/padlock-by-aj


Disclaimer
----------

THIS IS A PROOF-OF-CONCEPT.

It is ***NOT*** recommended that you store your two-factor authentication
secret keys on your hard-disk, as this COMPLETELY OBLITERATES most semblances
of security that two-factor authentication provides.  The whole point of
two-factor authentication is that a would-be attacker must jump through *two*
separate hoops:

1. (S)he must crack (or glean through hacking, social engineering, etc.) your
   password or passphrase to the system or service.
2. (S)he must gain access to the device with your two-factor secret key, which
   is usually your mobile phone, or a key fob which you should have on your
   person at all times.

Since it is likely that the computer you use to log into your other systems and
online services has its own form of password caching and/or storage, (unless
you’ve disabled it, which is a good idea if you’d like a bit of added security)
storing a second secret key somewhere on the machine nullifies this idea of key
separation, and makes it that much easier for a would-be attacker to gain
access to things they shouldn’t.

As it says in the license:

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE, TITLE AND NON-INFRINGEMENT. IN NO EVENT
SHALL THE COPYRIGHT HOLDERS OR ANYONE DISTRIBUTING THE SOFTWARE BE LIABLE
FOR ANY DAMAGES OR OTHER LIABILITY, WHETHER IN CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.

In other words, use this software — wisely, or unwisely — at YOUR OWN RISK.


License
-------

Boost Software License, Version 1.0: <http://www.boost.org/LICENSE_1_0.txt>


CHANGES.txt
v1.0.0, 2015-01-21 — First public version
v1.0.1, 2018-01-14 — Full Python 3.x support


test_hmac.py
"""Unit tests for Python's built-in hmac module."""

from unittest import TestCase

import hmac  # Unit Under Test


class HMACTestCase(TestCase):

    """
    These tests are taken straight out of RFC 2104.

    Python's hmac module also complies with the test vectors in RFC 4231.

    See Python's source code for more information.
    """

    def test_hmac_vector_1(self):
        key = '\x0b' * 16
        data = 'Hi There'
        digest = '9294727a3638bb1c13f48ef8158bfc9d'
        result = hmac.new(key, data).hexdigest()
        self.assertEqual(digest, result)

    def test_hmac_vector_2(self):
        key = 'Jefe'
        data = 'what do ya want for nothing?'
        digest = '750c783e6ab0b503eaa86e310a5db738'
        result = hmac.new(key, data).hexdigest()
        self.assertEqual(digest, result)

    def test_hmac_vector_3(self):
        key = '\xaa' * 16
        data = '\xdd' * 50
        digest = '56be34521d144c88dbb8c733f0e8b3f6'
        result = hmac.new(key, data).hexdigest()
        self.assertEqual(digest, result)


test_pyoath.py
# encoding=utf-8
"""Unit tests for the pyoath module."""

from __future__ import unicode_literals
import codecs
import hashlib
import struct
import sys
from io import StringIO
from unittest import TestCase

try:
    import builtins
except ImportError:  # Python <3
    import __builtin__ as builtins

from mock import mock_open, patch
from nose.tools import assert_dict_contains_subset, assert_equal, assert_raises

import pyoath  # Unit Under Test


def test_hmac_sha1():
    def test(i, expect):
        secret = b'12345678901234567890'
        count = struct.pack('>Q', i)[-8:]
        actual = pyoath._HMAC(secret, count)
        assert_equal(expect, actual)

    # These tests are taken straight out of RFC 4226.
    results = [
        codecs.decode('cc93cf18508d94934c64b65d8ba7667fb7cde4b0', 'hex'),
        codecs.decode('75a48a19d4cbe100644e8ac1397eea747a2d33ab', 'hex'),
        codecs.decode('0bacb7fa082fef30782211938bc1c5e70416ff44', 'hex'),
        codecs.decode('66c28227d03a2d5529262ff016a1e6ef76557ece', 'hex'),
        codecs.decode('a904c900a64b35909874b33e61c5938a8e15ed1c', 'hex'),
        codecs.decode('a37e783d7b7233c083d4f62926c7a25f238d0316', 'hex'),
        codecs.decode('bc9cd28561042c83f219324d3c607256c03272ae', 'hex'),
        codecs.decode('a4fb960c0bc06e1eabb804e5b397cdc4b45596fa', 'hex'),
        codecs.decode('1b3c89f65e6c9e883012052823443f048b4332db', 'hex'),
        codecs.decode('1637409809a679dc698207310c8c7fc07290d9e5', 'hex'),
    ]

    for i, expect in enumerate(results):
        yield test, i, expect


def test_DT():
    def test(i, expect):
        secret = b'12345678901234567890'
        count = struct.pack('>Q', i)[-8:]
        hmac_sha1 = pyoath._HMAC(secret, count)
        actual = pyoath._DT(hmac_sha1)
        assert_equal(expect, actual)

    # These tests are taken straight out of RFC 4226.
    results = [
        0x4c93cf18,
        0x41397eea,
        0x082fef30,
        0x66ef7655,
        0x61c5938a,
        0x33c083d4,
        0x7256c032,
        0x04e5b397,
        0x2823443f,
        0x2679dc69,
    ]

    for i, expect in enumerate(results):
        yield test, i, expect


def test_HOTP():
    def test(count, expect):
        secret = b'12345678901234567890'
        actual = pyoath.HOTP(secret, count)
        assert_equal(expect, actual)

    # These tests are taken straight out of RFC 4226.
    results = [
        '755224',
        '287082',
        '359152',
        '969429',
        '338314',
        '254676',
        '287922',
        '162583',
        '399871',
        '520489',
    ]

    for i, expect in enumerate(results):
        yield test, i, expect


def test_TOTP():
    @patch.object(pyoath, 'HOTP', wraps=pyoath.HOTP)
    @patch.object(pyoath.time, 'time')
    def test(unix_time, T, TOTP, Mode, mock_time, mock_HOTP):
        mock_time.return_value = unix_time

        # This is not noted in RFC 6238, but the secret changes with each Mode:
        # 20 bytes for SHA-1, 32 bytes for SHA-256, and 64 bytes for SHA-512.
        secret_len = len(Mode(b'x').digest())

        # The secret itself is just '0123456789',
        # repeating for as many bytes as necesary.
        secret = (b'1234567890' * 7)[:secret_len]

        result = pyoath.TOTP(secret, Digit=8, Mode=Mode)
        mock_HOTP.assert_called_once_with(secret, T, 8, Mode)
        assert_equal(TOTP, result)

    # These tests are taken straight out of RFC 6238.
    scenarios = [
        # 1970-01-01 00:00:59 Z
        (59, int('0000000000000001', 16), '94287082', hashlib.sha1),
        (59, int('0000000000000001', 16), '46119246', hashlib.sha256),
        (59, int('0000000000000001', 16), '90693936', hashlib.sha512),

        # 2005-03-18 01:58:29 Z
        (1111111109, int('00000000023523ec', 16), '07081804', hashlib.sha1),
        (1111111109, int('00000000023523ec', 16), '68084774', hashlib.sha256),
        (1111111109, int('00000000023523ec', 16), '25091201', hashlib.sha512),

        # 2005-03-18 01:58:31 Z
        (1111111111, int('00000000023523ed', 16), '14050471', hashlib.sha1),
        (1111111111, int('00000000023523ed', 16), '67062674', hashlib.sha256),
        (1111111111, int('00000000023523ed', 16), '99943326', hashlib.sha512),

        # 2009-02-13 23:31:30 Z
        (1234567890, int('000000000273ef07', 16), '89005924', hashlib.sha1),
        (1234567890, int('000000000273ef07', 16), '91819424', hashlib.sha256),
        (1234567890, int('000000000273ef07', 16), '93441116', hashlib.sha512),

        # 2033-05-18 03:33:20 Z
        (2000000000, int('0000000003f940aa', 16), '69279037', hashlib.sha1),
        (2000000000, int('0000000003f940aa', 16), '90698825', hashlib.sha256),
        (2000000000, int('0000000003f940aa', 16), '38618901', hashlib.sha512),

        # 2603-10-11 11:33:20 Z
        (20000000000, int('0000000027bc86aa', 16), '65353130', hashlib.sha1),
        (20000000000, int('0000000027bc86aa', 16), '77737706', hashlib.sha256),
        (20000000000, int('0000000027bc86aa', 16), '47863826', hashlib.sha512),
    ]

    for unix_time, T, TOTP, Mode, in scenarios:
        yield test, unix_time, T, TOTP, Mode


def test_get_chmod_bits():
    def test(supply, expect):
        r, w, x = expect
        bits = pyoath._get_chmod_bits(supply)
        assert_equal(r, (bits >> 8) & 1, 'read permissions mis-match')
        assert_equal(w, (bits >> 7) & 1, 'write permissions mis-match')
        assert_equal(x, (bits >> 6) & 1, 'execute permissions mis-match')

    scenarios = [
        ('pyoath.py', (1, 1, 1)),
        ('test_pyoath.py', (1, 1, 0)),
    ]

    for supply, expect in scenarios:
        yield test, supply, expect


def test_graceful_encode():
    def test(supply, encoding, expect):
        if isinstance(expect, type) and issubclass(expect, BaseException):
            assert_raises(expect, pyoath._graceful_encode, supply, encoding)
        else:
            actual = pyoath._graceful_encode(supply, encoding)
            assert_equal(expect, actual)

    scenarios = [(
        'ASCII, ISO-8859-1 (LATIN-1), and WINDOWS-1252 are dead.', 'utf-8',
        b'ASCII, ISO-8859-1 (LATIN-1), and WINDOWS-1252 are dead.',
    ), (
        b'ASCII, ISO-8859-1 (LATIN-1), and WINDOWS-1252 are dead.', 'utf-8',
        b'ASCII, ISO-8859-1 (LATIN-1), and WINDOWS-1252 are dead.',
    ), (
        'Long live Üñîçø∂é and ü†ƒ-∞!', 'utf-8',
        'Long live Üñîçø∂é and ü†ƒ-∞!'.encode('utf-8'),
    ), (
        'Long live Üñîçø∂é and ü†ƒ-∞!'.encode('utf-8'), 'utf-8',
        'Long live Üñîçø∂é and ü†ƒ-∞!'.encode('utf-8'),
    ), (
        b'\x4a\xc0\xff\xee', 'utf-8',  # Raw binary comes through unharmed
        b'\x4a\xc0\xff\xee',
    ), (
        '€ ¢ £ ¤ ¥', 'windows-1252',  # WINDOWS-1252 is commonly mistaken for
        b'\x80 \xa2 \xa3 \xa4 \xa5',  # LATIN-1, but LATIN-1 does not have Euro
    ), (
        '€ ¢ £ ¤ ¥', 'iso-8859-1',  # ISO-8859-1, aka LATIN-1
        SystemExit,
    ), (
        Ellipsis, '…',  # Unsupported object type
        Exception,
    )]

    for supply, encoding, expect in scenarios:
        yield test, supply, encoding, expect


def test_parse_args():
    def test(supply, expect):
        actual = dict(vars(pyoath._parse_args(supply)))
        assert_dict_contains_subset(expect, actual)

    scenarios = [(
        ['opensesame'],
        {'google': False, 'loop': False, 'secret': 'opensesame'},
    ), (
        ['N5YGK3TTMVZWC3LF', '--google'],
        {'google': True, 'loop': False, 'secret': 'N5YGK3TTMVZWC3LF'},
    ), (
        ['opensesame', '--loop'],
        {'google': False, 'loop': True, 'secret': 'opensesame'},
    ), (
        ['N5YGK3TTMVZWC3LF', '--google', '--loop'],
        {'google': True, 'loop': True, 'secret': 'N5YGK3TTMVZWC3LF'},
    )]

    for supply, expect in scenarios:
        yield test, supply, expect


class MainTest(TestCase):
    def setUp(self):
        patch.object(
            pyoath, 'expanduser', new=lambda s: s.replace('~', '/home/user', 1)
        ).start()

        patch.object(
            pyoath, 'realpath', new=lambda s: ('/abs/' + s).replace('//', '/')
        ).start()

        self.mock_chmod = patch.object(pyoath, '_get_chmod_bits').start()
        self.mock_isfile = patch.object(pyoath, 'isfile').start()
        self.mock_sleep = patch.object(pyoath.time, 'sleep').start()
        self.mock_time = patch.object(pyoath.time, 'time').start()

        self.mock_time.return_value = 1111111109  # 2005-03-18 01:58:29 Z

        self.mock_open = mock_open()
        self.mock_file = self.mock_open.return_value

        self.stdout = StringIO()
        self.stderr = StringIO()
        patch.object(pyoath.sys, 'stdout', new=self.stdout).start()
        patch.object(pyoath.sys, 'stderr', new=self.stderr).start()

    def tearDown(self):
        patch.stopall()

    def _args(self, *args):
        if sys.version_info >= (3, 0, 0):
            return args
        else:
            return map(lambda s: s.encode('utf-8'), args)

    def _assert_stdout(self, contents):
        self.stdout.seek(0)
        self.assertEqual(contents, self.stdout.read())

    def _assert_stderr(self, contents):
        self.stderr.seek(0)
        self.assertEqual(contents, self.stderr.read())

    def test_happy_path(self):
        self.mock_isfile.return_value = False
        result = pyoath.main(*self._args('12345678901234567890'))
        assert_equal(0, result)
        self.mock_isfile.assert_called_once_with('/abs/12345678901234567890')
        self._assert_stdout('081804\n')
        self._assert_stderr('')

    def test_looping_authenticator_opened_on_very_first_second(self):
        self.mock_isfile.return_value = False
        self.mock_time.side_effect = list(
            range(1111111110 - 30, 1111111110 + 4)
        ) + [KeyboardInterrupt]
        result = pyoath.main(*self._args(
            '12345678901234567890', '--loop'
        ))
        assert_equal(0, result)
        self._assert_stdout("""
Authenticator Started!
:----------------------------:--------:
:       Code Wait Time       :  Code  :
:----------------------------:--------:
+++++++++++++++++++++++++++++: 081804 :
.............................: 050471 :
...""")
        self._assert_stderr('')

    def test_looping_authenticator_opened_on_second_second(self):
        self.mock_isfile.return_value = False
        self.mock_time.side_effect = list(
            range(1111111110 - 29, 1111111110 + 4)
        ) + [KeyboardInterrupt]
        result = pyoath.main(*self._args(
            '12345678901234567890', '--loop'
        ))
        assert_equal(0, result)
        self._assert_stdout("""
Authenticator Started!
:----------------------------:--------:
:       Code Wait Time       :  Code  :
:----------------------------:--------:
+++++++++++++++++++++++++++++: 081804 :
............................+: 050471 :
...""")
        self._assert_stderr('')

    def test_looping_authenticator_opened_on_fifteenth_second(self):
        self.mock_isfile.return_value = False
        self.mock_time.side_effect = list(
            range(1111111110 - 16, 1111111110 + 4)
        ) + [KeyboardInterrupt]
        result = pyoath.main(*self._args(
            '12345678901234567890', '--loop'
        ))
        assert_equal(0, result)
        self._assert_stdout("""
Authenticator Started!
:----------------------------:--------:
:       Code Wait Time       :  Code  :
:----------------------------:--------:
+++++++++++++++++++++++++++++: 081804 :
...............++++++++++++++: 050471 :
...""")
        self._assert_stderr('')

    def test_looping_authenticator_opened_on_twenty_ninth_second(self):
        self.mock_isfile.return_value = False
        self.mock_time.side_effect = list(
            range(1111111110 - 2, 1111111110 + 4)
        ) + [KeyboardInterrupt]
        result = pyoath.main(*self._args(
            '12345678901234567890', '--loop'
        ))
        assert_equal(0, result)
        self._assert_stdout("""
Authenticator Started!
:----------------------------:--------:
:       Code Wait Time       :  Code  :
:----------------------------:--------:
+++++++++++++++++++++++++++++: 081804 :
.++++++++++++++++++++++++++++: 050471 :
...""")
        self._assert_stderr('')

    def test_looping_authenticator_opened_on_very_last_second(self):
        self.mock_isfile.return_value = False
        self.mock_time.side_effect = [
            1111111109,
            1111111110,
            1111111111,
            1111111112,
            1111111113,
            KeyboardInterrupt,
        ]
        result = pyoath.main(*self._args(
            '12345678901234567890', '--loop'
        ))
        assert_equal(0, result)
        self._assert_stdout("""
Authenticator Started!
:----------------------------:--------:
:       Code Wait Time       :  Code  :
:----------------------------:--------:
+++++++++++++++++++++++++++++: 081804 :
+++++++++++++++++++++++++++++: 050471 :
...""")
        self._assert_stderr('')

    def test_read_secret_from_raw_2fa_file(self):
        self.mock_isfile.return_value = True
        self.mock_chmod.return_value = 0o600
        self.mock_file.read.return_value = b'12345678901234567890'
        with patch.object(builtins, 'open', self.mock_open):
            result = pyoath.main(*self._args('secret.2fa'))
        assert_equal(0, result)
        self.mock_open.assert_called_once_with('/abs/secret.2fa', 'rb')
        self.mock_file.read.assert_called_once_with()
        self._assert_stdout('081804\n')
        self._assert_stderr('')

    def test_read_secret_from_google_2fa_file(self):
        self.mock_isfile.return_value = True
        self.mock_chmod.return_value = 0o600
        self.mock_file.read.return_value = b'GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ'
        with patch.object(builtins, 'open', self.mock_open):
            result = pyoath.main(*self._args('google.2fa', '--google'))
        assert_equal(0, result)
        self.mock_open.assert_called_once_with('/abs/google.2fa', 'rb')
        self.mock_file.read.assert_called_once_with()
        self._assert_stdout('081804\n')
        self._assert_stderr('')

    def test_refuse_to_read_unprotected_2fa_file(self):
        self.mock_isfile.return_value = True
        self.mock_chmod.return_value = 0o644
        with patch.object(builtins, 'open', self.mock_open):
            result = pyoath.main(*self._args('secret.2fa'))
        assert_equal(2, result)
        self.mock_chmod.assert_called_once_with('/abs/secret.2fa')
        self._assert_stdout('')
        self._assert_stderr("""\
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@          WARNING: UNPROTECTED SECRET 2FA FILE!          @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
Permissions 0644 for '/abs/secret.2fa' are too open.
It is required that your secret key files are NOT accessible by others.
This secret key will be ignored.
""")


test_builtins.py
"""Unit tests for Python's built-in hashlib and hmac modules."""

import hashlib  # Unit Under Test
import hmac  # Unit Under Test

from nose.tools import assert_equal


def test_hashlib_sha1():
    def test(supply, factor, expect):
        actual = hashlib.sha1(supply * factor).hexdigest()
        assert_equal(expect, actual)

    # These tests are taken straight out of RFC 3174.
    scenarios = [(
        b'abc', 1,
        'a9993e364706816aba3e25717850c26c9cd0d89d',
    ), (
        b'abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq', 1,
        '84983e441c3bd26ebaae4aa1f95129e5e54670f1',
    ), (
        b'a', 1000000,
        '34aa973cd4c4daa4f61eeb2bdbad27316534016f',
    ), (
        b'01234567012345670123456701234567', 20,
        'dea356a2cddd90c7a7ecedc5ebb563934f460452',
    )]

    for supply, factor, expect in scenarios:
        yield test, supply, factor, expect


def test_hmac_md5():
    def test(key, factor_1, msg, factor_2, expect):
        actual = hmac.new(key * factor_1, msg * factor_2).hexdigest()
        assert_equal(expect, actual)

    # These tests are taken straight out of RFC 2104.
    scenarios = [(
        b'\x0b', 16,
        b'Hi There', 1,
        '9294727a3638bb1c13f48ef8158bfc9d',
    ), (
        b'Jefe', 1,
        b'what do ya want for nothing?', 1,
        '750c783e6ab0b503eaa86e310a5db738',
    ), (
        b'\xaa', 16,
        b'\xdd', 50,
        '56be34521d144c88dbb8c733f0e8b3f6',
    )]

    for key, factor_1, msg, factor_2, expect in scenarios:
        yield test, key, factor_1, msg, factor_2, expect


pyoath.py
#! /usr/bin/env python
# encoding=utf-8

"""
A Python OATH implementation.

OATH is the Initiative for Open Authentication - not to be confused with OAuth,
the Open Standard to *Authorization*, which is an entirely different paradigm.

PyOATH implements the HOTP Algorithm defined in RFC 4226, published in December
of 2005, and the TOTP Algorithm defined in RFC 6238, published in May of 2011.
It has been designed for both the client- and server-sides of two-factor
authentication systems.

Many of the variable names in this code were used because those were the names
used in those publications. Much of the documentation and many of the comments
in this file was taken straight from the RFC documentation as well.

Please read RFC 4226 and RFC 6238 for more information.
"""

from __future__ import unicode_literals
import argparse
import base64
import hashlib
import hmac
import os
import stat
import struct
import sys
import time
from os.path import expanduser, isfile, realpath


def _DT(String):
    """
    Dynamic Truncation of an HMAC-SHA-1 value.

        DT(String) // String = String[0]...String[19]
            Let OffsetBits be the low-order 4 bits of String[19]
            Offset = StToNum(OffsetBits) // 0 <= OffSet <= 15
            Let P = String[OffSet]...String[OffSet+3]
            Return the Last 31 bits of P

    :param String: An HMAC value. If HMAC-SHA-1, it's 160-bits (20-bytes) long.
    :type String: bytes
    :return: The 31-bit dynamically truncated HMAC value
    :rtype: int
    """
    OffsetBits = String[-1::1]
    Offset = ord(OffsetBits) & 0x0f
    P = String[Offset:Offset+4]
    Bits = _StToNum(P)

    # The reason for masking the most significant bit of P is to avoid
    # confusion about signed vs. unsigned modulo computations.  Different
    # processors perform these operations differently, and masking out the
    # signed bit removes all ambiguity.
    Bits &= 0x7fffffff

    return Bits


def _HMAC(K, C, Mode=hashlib.sha1):
    """
    Generate an HMAC value.

    The default mode is to generate an HMAC-SHA-1 value w/ the SHA-1 algorithm.

    :param K: shared secret between client and server.
            Each HOTP generator has a different and unique secret K.
    :type K: bytes
    :param C: 8-byte counter value, the moving factor.
            This counter MUST be synchronized between the HOTP generator
            (client) and the HOTP validator (server).
    :type C: bytes
    :param Mode: The algorithm to use when generating the HMAC value
    :type Mode: hashlib.sha1, hashlib.sha256, hashlib.sha512, or hashlib.md5
    :return: HMAC result. If HMAC-SHA-1, result is 160-bits (20-bytes) long.
    :rtype: bytes
    """
    return hmac.new(K, C, Mode).digest()


def _StToNum(S):
    """
    Convert S to a number.

    :param S: The (big-endian) bytestring to convert to an integer
    :type S: bytes
    :return: An integer representation of the bytestring (rightmost chr == LSB)
    :rtype: int
    """
    return struct.unpack('>L', S)[0]


def _Truncate(HS, Digit=6):
    """
    Convert an HMAC value into an HOTP value.

    NOTE:
    Implementations MUST extract a 6-digit code at a minimum and possibly
    7 and 8-digit code. Depending on security requirements, Digit = 7 or
    more SHOULD be considered in order to extract a longer HOTP value.

    :param HS: An HMAC bytestring. If HMAC-SHA-1, is 160-bits (20-bytes) long.
    :type HS: bytes
    :param Digit: Digits to extract from the dynamically truncated HMAC value
    :type Digit: int
    :return: The final HOTP value
    :rtype: str
    """
    # The Truncate function performs Step 2 and Step 3, i.e., the dynamic
    # truncation and then the reduction modulo 10^Digit. The purpose of
    # the dynamic offset truncation technique is to extract a 4-byte
    # dynamic binary code from a 160-bit (20-byte) HMAC-SHA-1 result.
    Snum = _DT(HS)  # Step 2+3
    D = Snum % (10 ** Digit)
    return str(D).zfill(Digit)


def HOTP(K, C, Digit=6, Mode=hashlib.sha1):
    """
    HOTP: An HMAC-Based One-Time Password Algorithm.

    The algorithm for this function is defined in RFC 4226, Section 5.3.

    :param K: shared secret between client and server.
            Each HOTP generator has a different and unique secret K.
    :type K: bytes
    :param C: the counter value (as an int), the moving factor.
            This counter MUST be synchronized between the HOTP generator
            (client) and the HOTP validator (server).
    :type C: int
    :param Digit: Digits to extract from the dynamically truncated HMAC value
    :type Digit: int
    :param Mode: The algorithm to use when generating the HMAC value
    :type Mode: hashlib.sha1, hashlib.sha256, hashlib.sha512, or hashlib.md5
    :return: An HMAC-Based One-Time Password
    :rtype: str
    """
    # We can describe the operations in 3 distinct steps:
    # Step 1: Generate an HMAC value
    C_bytestr = struct.pack('>Q', C)[-8:]  # Pack int C into an 8-byte string
    HS = _HMAC(K, C_bytestr, Mode)  # HS is a >= 20-byte string (if using SHA1)

    # Step 2: Generate a 4-byte string (Dynamic Truncation)
    # Step 3: Compute an HOTP value
    return _Truncate(HS, Digit)


def TOTP(K, X=30, Digit=6, Mode=hashlib.sha1):
    """
    TOTP: Time-Based One-Time Password Algorithm.

    This variant of the HOTP algorithm specifies the calculation of a
    one-time password value, based on a representation of the counter as
    a time factor.

    The algorithm for this function is defined in RFC 6238, Section 4.2.

    NOTE:
    TOTP implementations MAY use HMAC-SHA-256 or HMAC-SHA-512 functions,
    based on SHA-256 or SHA-512 [SHA2] hash functions, instead of the
    HMAC-SHA-1 function that has been specified for the HOTP computation
    in [RFC4226].

    :param K: shared secret between client and server.
            Each HOTP generator has a different and unique secret K.
    :type K: bytes
    :param X: represents the time step in seconds (default value X = 30)
    :type X: int
    :param Digit: Digits to extract from the dynamically truncated HMAC value
    :type Digit: int
    :param Mode: The algorithm to use when generating the HMAC value
    :type Mode: hashlib.sha1, hashlib.sha256, hashlib.sha512, or hashlib.md5
    :return: A Time-Based One-Time Password
    :rtype: str
    """
    unix_time = int(time.time())
    unix_step = int(unix_time / X)
    return HOTP(K, unix_step, Digit, Mode)


def _get_chmod_bits(file_path):
    """Get the CHMOD bits of a file."""
    s = os.stat(file_path)
    return s.st_mode & (stat.S_IRWXU | stat.S_IRWXG | stat.S_IRWXO)


def _graceful_encode(s, encoding='utf-8', errors='strict'):
    """
    Convert a unicode string to a UTF-8 encoded bytestring.

    If the string is already a bytestring, return it unmodified.

    :param s: The string to encode as a bytestring
    :type s: str
    :param encoding: The encoding to use. Defaults to UTF-8.
    :type encoding: str
    :param errors: The error policy to employ when lossless encoding fails.
        Defaults to strict, which raises and exception. Can be set to replace
        for lossy encodes.
    :type errors: str
    :return: The string in encoded bytestring form,
            or the original input if the string was already a bytestring.
    :rtype: bytes
    """
    try:
        return s.encode(encoding=encoding, errors=errors)
    except (AttributeError, UnicodeDecodeError):
        if hasattr(s, 'startswith'):  # Already a bytestring
            return s
        raise
    except UnicodeEncodeError:
        sys.stderr.write("""\
Python 3 does not support arbitrary binary input supplied as command-line args.
If your key contains raw binary, write it to a file, protect its permissions
with, e.g., `chmod 600 secret.2fa`, and then pass the file path as an argument
to pyoath.
""")
        sys.exit(1)


def _parse_args(args):
    """Parse command-line arguments."""
    parser = argparse.ArgumentParser(description=__doc__)
    arg = parser.add_argument

    arg('secret',
        help='shared secret file between client and server',
        type=str)

    arg('--google',
        help='Google Authenticator mode (assumes secret is encoded in base32)',
        action='store_true',
        required=False)

    arg('--loop',
        help='start an authenticator instance that will continue until killed',
        action='store_true',
        required=False)

    return parser.parse_args(args or sys.argv[1:])


def main(*args):
    """Program entry point for command-line use."""
    options = _parse_args(args)
    secret = options.secret
    secret_path = realpath(expanduser(secret))

    if isfile(secret_path):
        # Encourage good old-fashioned practices
        # with some good old-fashioned butt-kicking.
        mode = _get_chmod_bits(secret_path)
        if mode & 0o077 != 0:
            msg = """\
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@          WARNING: UNPROTECTED SECRET 2FA FILE!          @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
Permissions 0%o for '%s' are too open.
It is required that your secret key files are NOT accessible by others.
This secret key will be ignored.
""" % (mode, secret_path)
            sys.stderr.write(msg)
            return 2

        with open(secret_path, 'rb') as f:
            secret = f.read()
            if options.google:
                contents = secret.decode('utf-8').strip().upper()
                secret = base64.b32decode(contents)

    secret = _graceful_encode(secret)

    if options.loop:
        last_otp = None
        counter = 30
        sys.stdout.write("""
Authenticator Started!
:----------------------------:--------:
:       Code Wait Time       :  Code  :
:----------------------------:--------:
""")
        while True:
            try:
                this_otp = TOTP(secret)
                counter -= 1
                if this_otp == last_otp:
                    sys.stdout.write('.')
                else:
                    last_otp = this_otp
                    sys.stdout.write('+' * counter + ': %s :\n' % this_otp)
                    counter = 30
                sys.stdout.flush()
                time.sleep(1)
            except KeyboardInterrupt:
                return 0

    otp = TOTP(secret)
    sys.stdout.write(otp + '\n')
    return 0


if __name__ == '__main__':
    sys.exit(main(*sys.argv[1:]))
